diffusion/models/autoencoder.py

```diff
@@ -307,32 +356,54 @@ def __init__(self,
         self.post_quant_conv = nn.Conv2d(self.latent_channels, self.latent_channels, kernel_size=1, stride=1, padding=0)
         nn.init.kaiming_normal_(self.post_quant_conv.weight, nonlinearity='linear')
 
+    @property
+    def device(self) -> torch.device:
+        return next(self.parameters()).device
+
+    def get_extra_state(self):
+        return {'config': self.config}
+
+    def set_extra_state(self, state):
+        # Save the autoencoder config
+        config = {}
+        config['input_channels'] = self.input_channels
+        config['output_channels'] = self.output_channels
+        config['hidden_channels'] = self.hidden_channels
+        config['latent_channels'] = self.latent_channels
+        config['double_latent_channels'] = self.double_latent_channels
+        config['channel_multipliers'] = self.channel_multipliers
+        config['num_residual_blocks'] = self.num_residual_blocks
+        config['use_conv_shortcut'] = self.use_conv_shortcut
+        config['dropout_probability'] = self.dropout_probability
+        config['resample_with_conv'] = self.resample_with_conv
+        config['use_attention'] = self.use_attention
+        config['zero_init_last'] = self.zero_init_last
+        self.config = config
+
     def get_last_layer_weight(self) -> torch.Tensor:
         """Get the weight of the last layer of the decoder."""
         return self.decoder.conv_out.weight
 
-    def encode(self, x: torch.Tensor) -> Dict[str, torch.Tensor]:
+    def encode(self, x: torch.Tensor) -> GaussianDistribution:
         """Encode an input tensor into a latent tensor."""
         h = self.encoder(x)
         moments = self.quant_conv(h)
         # Split the moments into mean and log variance
         mean, log_var = moments[:, :self.latent_channels], moments[:, self.latent_channels:]
-        return {'mean': mean, 'log_var': log_var}
+        return GaussianDistribution(mean, log_var)
 
-    def decode(self, z: torch.Tensor) -> Dict[str, torch.Tensor]:
+    def decode(self, z: torch.Tensor) -> AutoEncoderOutput:
         """Decode a latent tensor into an output tensor."""
         z = self.post_quant_conv(z)
         x_recon = self.decoder(z)
-        return {'x_recon': x_recon}
+        return AutoEncoderOutput(x_recon)
```

Comment:
Yes, they do. Do you think it's needed here? I was aiming for the minimal implementation that's compatible with our current model classes rather than compatibility with HF. A cleaner option could be to change our model classes to use the HF `return_dict` and then switch this back to dict outputs.